**NOMBRE ALUMNO:** Jonathan Rodriguez

**FECHA:** 15-10-19

## Desafío 1: Lectura de archivos

* Importe pandas y numpy siguiendo las convenciones.
* Lea la base de datos worldcup2014.csv y asígnela a un objeto df .
* Solicite las primeras 5 observaciones con head

In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv('worldcup2014.csv')

df.head(5)

team     continent group  group_pos  cantidad_juegos  \
0       Brazil  southamerica     A          1                3   
1       Mexico  northamerica     A          2                3   
2      Croatia        europe     A          3                3   
3     Cameroon        africa     A          4                3   
4  Netherlands        europe     B          1                3   

   juegos_ganados  juegos_empatados  juegos_perdidos  goles_favor  \
0               2                 1                0            7   
1               2                 1                0            4   
2               1                 0                2            6   
3               0                 0                3            1   
4               3                 0                0           10   

   goles_contra goles_diferencia  puntos  clasificado  
0             2                5       7            1  
1             1                3       7            1  
2             6                0       3            0  
3             9               −8       0            0  
4             3                7       9            1

## Desafío 2: Estime las frecuencias de continent

Utilice value_counts . Responda lo siguiente:

* ¿Cuál es el continente con una mayor presencia en la muestra?

In [2]:
contiente_con_mayor_presencia_por_continente = df['continent'].value_counts().head(1)
contiente_con_mayor_presencia_por_continente

europe    13
Name: continent, dtype: int64

_El continente con mayor presencia en la muestra es europa con 13 países_

* ¿Cuál es la probabilidad de elegir un equipo asiático al azar?


In [3]:
cantidad_por_contiente = df['continent'].value_counts('%')
cantidad_por_contiente

europe          0.40625
southamerica    0.18750
africa          0.15625
asia            0.12500
northamerica    0.12500
Name: continent, dtype: float64

La probabilidad de elegir un equipo asiático al azar en función del resultado es de 0.125

* ¿Cuál es la probabilidad de elegir un equipo africano al azar?

En función de la variable anterior, la probabilidad de elegir un equipo africano al azar es de 0.15625

## Desafío 3: Probabilidades por continente

Por cada continente, genere un nuevo objeto que almacene sólo las observaciones del continente.

_tip: Para ello puede utilizar la siguiente sintáxis: df[df['variable'] == condicion]_

In [4]:
dfEurope = df[df['continent'] == 'europe']
dfAsia = df[df['continent'] == 'asia']
dfAfrica = df[df['continent'] == 'africa']
dfNorthAmerica = df[df['continent'] == 'northamerica']
dfSouthAmerica = df[df['continent'] == 'southamerica']


## Desafío 4: Calcule la frecuencia

Pase todos los objetos creados en el desafio anterior en un loop que imprima la frecuencia con la que pasaron a la siguiente ronda

_tip: Genere un array que contenga los objetos a pasar en el loop._

In [108]:
cantidad_paises = len(df)
arreglo_clasificar_region = []
arreglo_probabilidad_clasificados_continente = []
arreglo_frecuencia_clasificados_continente = [] 
for continente in [dfEurope, dfAsia, dfAfrica, dfNorthAmerica, dfSouthAmerica]:
    contador_region = len(continente)
    contador_clasificados = len(continente[continente['clasificado'] == 1])
    probabilidad_clasificados = contador_clasificados/contador_region
    arreglo_probabilidad_clasificados_continente.append(probabilidad_clasificados)
    arreglo_frecuencia_clasificados_continente.append(contador_clasificados)
    
print('Frecuencia absoluta de clasificación por continente')
frecuencia_clasificar = dict(zip(["europe", "asia", "africa", "northamerica", "southamerica"], arreglo_frecuencia_clasificados_continente))        
print(frecuencia_clasificar)
print('\n')
print('Frecuencia relativa de clasificación por continente')
probabilidad_clasificar = dict(zip(["europe", "asia", "africa", "northamerica", "southamerica"], arreglo_probabilidad_clasificados_continente))        
print(probabilidad_clasificar)





Frecuencia absoluta de clasificación por continente
{'europe': 6, 'asia': 0, 'africa': 2, 'northamerica': 3, 'southamerica': 5}


Frecuencia relativa de clasificación por continente
{'europe': 0.46153846153846156, 'asia': 0.0, 'africa': 0.4, 'northamerica': 0.75, 'southamerica': 0.8333333333333334}


_El diccionario anterior nos muestra la frecuencia y probabilidad de clasificación en función de cada continente_

* ¿Cuál fue la frecuencia de que un país asiático pase a la siguiente ronda?

In [72]:
print('Frecuencia absoluta')
print(frecuencia_clasificar['asia'])
print('Frecuencia relativa')
print(probabilidad_clasificar['asia'])


Frecuencia absoluta
0
Frecuencia relativa
0.0


_La frecuencia de clasifación de un país asiático fue de 0_

* ¿Cuáles fueron los dos continentes que tuvieron la mayor frecuencia de clasificar?

In [84]:
organizar = sorted(frecuencia_clasificar.items(), key=lambda kv: kv[1], reverse=True)
print(organizar[0])
print(organizar[1])

('europe', 6)
('southamerica', 5)


_El resultado muestra aquellos dos continentes con mayor frecuencia absoluta_

In [73]:
organizar = sorted(probabilidad_clasificar.items(), key=lambda kv: kv[1], reverse=True)
print(organizar[0])
print(organizar[1])

('southamerica', 0.8333333333333334)
('northamerica', 0.75)


_El resultado muestra aquellos dos continentes con mayor frecuencia relativa_

* ¿Cuál fue la probabilidad de no clasificar para el continente europeo?

In [26]:
probabilibad_clasificar_europe = probabilidad_clasificar['europe']
probabilidad_no_clasificar = 1 - probabilibad_clasificar_europe
print(probabilidad_no_clasificar)


no_clasifica = len(dfEurope[dfEurope['clasificado'] == 0])
probabilidad_no_clasificado = no_clasifica/len(dfEurope)
print(probabilidad_no_clasificado)

0.5384615384615384
0.5384615384615384


* ¿Cuál fue la frecuencia de no clasificar un país europeo?

In [59]:
no_clasifica = len(dfEurope[dfEurope['clasificado'] == 0])
total_europa = len(dfEurope) 
print("La frecuencia absoluta de los paises europeos que no clasificaron fue de {} de un total de {} ".format(no_clasifica, total_europa))


La frecuencia absoluta de los paises europeos que no clasificaron fue de 7 de un total de 13 


_La probabilidad de no clasificar de un país europeo es de: 0.538_

## Desafío 5: Refactorización

* A continuación se presenta un loop que cuenta las probabilidades de juegos ganados. Se pide que value_counts cuente los juegos no ganados y ganados, ignorando de la cantidad de juegos.


```
for d in [africa_df, europe_df, asia_df, northamerica_df, southamerica_df]:
   print(d['juegos_ganados'].value_counts('%'))

```


### Para ello puede utilizar np.where , que devuelve un array con valores imputados cuando se satisface una condición. En este caso debemos clasificar como 0 cuando los juegos ganados sean igual a 0, y 1 cuando sean igual o mayor a 1.


In [142]:
arreglo_ganados = []
arreglo_frecuencia_ganados = []
arreglo_frecuencia_absoluta_ganados = []
arreglo_cantidad_total_juegos_ganados = []
arreglo_frecuencia_relativa = []
arreglo_frecuencia_en_funcion_total = []
for continente in [dfEurope, dfAsia, dfAfrica, dfNorthAmerica, dfSouthAmerica]:
    continente['ganados'] = np.where(continente['juegos_ganados'] == 0, 0, 1)
    cantidad_ganados = len(continente[continente['ganados'] == 1])
    cantidad_paises_continente = len(continente)
    calculo_frecuencia_ganados = cantidad_ganados/cantidad_paises_continente
    arreglo_ganados.append(cantidad_ganados)
    arreglo_frecuencia_ganados.append(calculo_frecuencia_ganados)
    arreglo_frecuencia_absoluta_ganados.append(continente['juegos_ganados'].sum())
    arreglo_cantidad_total_juegos_ganados.append(continente['cantidad_juegos'].sum())
    arreglo_frecuencia_en_funcion_total.append(continente['juegos_ganados'].sum()/continente['cantidad_juegos'].sum())

print('Cantidad de juegos ganados frecuencia absoluta')    
frecuencia_absoluta_ganados = dict(zip(["europe", "asia", "africa", "northamerica", "southamerica"], arreglo_frecuencia_absoluta_ganados))        
print(frecuencia_absoluta_ganados)   
print('\n')
print('Cantidad de juegos')    
cantidad_total_juegos_ganados = dict(zip(["europe", "asia", "africa", "northamerica", "southamerica"], arreglo_cantidad_total_juegos_ganados))        
print(cantidad_total_juegos_ganados)   
print('\n') 
print('Frecuencia relativa en función del total de juegos')
frecuencia_en_funcion_total_ganados = dict(zip(["europe", "asia", "africa", "northamerica", "southamerica"], arreglo_frecuencia_en_funcion_total))        
print(frecuencia_en_funcion_total_ganados)   
print('\n') 
print('Frecuencia utilizando la lógica del enunciado con np.where')
frecuencia_ganados = dict(zip(["europe", "asia", "africa", "northamerica", "southamerica"], arreglo_frecuencia_ganados))        
print(frecuencia_ganados)

Cantidad de juegos ganados frecuencia absoluta
{'europe': 18, 'asia': 0, 'africa': 3, 'northamerica': 5, 'southamerica': 13}


Cantidad de juegos
{'europe': 39, 'asia': 12, 'africa': 15, 'northamerica': 12, 'southamerica': 18}


Frecuencia relativa en función del total de juegos
{'europe': 0.46153846153846156, 'asia': 0.0, 'africa': 0.2, 'northamerica': 0.4166666666666667, 'southamerica': 0.7222222222222222}


Frecuencia utilizando la lógica del enunciado con np.where
{'europe': 0.8461538461538461, 'asia': 0.0, 'africa': 0.6, 'northamerica': 0.75, 'southamerica': 1.0}


/home/jonathan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


* ¿Qué continente tuvo mayor frecuencia de juegos ganados?


In [143]:
print('Frecuencia absoluta en función del total de juegos ganados')
frecuencia_juegos_frecuencia_absoluta = sorted(frecuencia_absoluta_ganados.items(), key=lambda kv: kv[1], reverse=True)
print(frecuencia_juegos_frecuencia_absoluta[0])
print('\n')
print('Frecuencia relativa de juegos ganados en función del total de juegos ganados')
frecuencia_juegos_frecuencia_relativa = sorted(frecuencia_en_funcion_total_ganados.items(), key=lambda kv: kv[1], reverse=True)
print(frecuencia_juegos_frecuencia_relativa[0])
print('\n')
print('Frecuencia de juegos ganados en función del enunciado con planteamiento de np.where')
frecuencia_juegos = sorted(frecuencia_ganados.items(), key=lambda kv: kv[1], reverse=True)
print(frecuencia_juegos[0])

Frecuencia absoluta en función del total de juegos ganados
('europe', 18)


Frecuencia relativa de juegos ganados en función del total de juegos ganados
('southamerica', 0.7222222222222222)


Frecuencia de juegos ganados en función del enunciado con planteamiento de np.where
('southamerica', 1.0)


In [144]:
arreglo_perdidos = []
arreglo_frecuencia_perdidos = []
arreglo_frecuencia_absoluta_perdidos = []
arreglo_cantidad_total_juegos_perdidos = []
arreglo_frecuencia_relativa = []
arreglo_frecuencia_en_funcion_total = []
for continente in [dfEurope, dfAsia, dfAfrica, dfNorthAmerica, dfSouthAmerica]:
    continente['perdidos'] = np.where(continente['juegos_perdidos'] == 0, 0, 1)
    cantidad_perdidos = len(continente[continente['perdidos'] == 1])
    cantidad_paises_continente = len(continente)
    calculo_frecuencia_perdidos = cantidad_perdidos/cantidad_paises_continente
    arreglo_perdidos.append(cantidad_perdidos)
    arreglo_frecuencia_perdidos.append(calculo_frecuencia_perdidos)
    arreglo_frecuencia_absoluta_perdidos.append(continente['juegos_perdidos'].sum())
    arreglo_cantidad_total_juegos_perdidos.append(continente['cantidad_juegos'].sum())
    arreglo_frecuencia_en_funcion_total.append(continente['juegos_perdidos'].sum()/continente['cantidad_juegos'].sum())

print('Cantidad de juegos perdidos frecuencia absoluta')    
frecuencia_absoluta_perdidos = dict(zip(["europe", "asia", "africa", "northamerica", "southamerica"], arreglo_frecuencia_absoluta_perdidos))        
print(frecuencia_absoluta_perdidos)   
print('\n')
print('Cantidad de juegos')    
cantidad_total_juegos_perdidos = dict(zip(["europe", "asia", "africa", "northamerica", "southamerica"], arreglo_cantidad_total_juegos_perdidos))        
print(cantidad_total_juegos_perdidos)   
print('\n') 
print('Frecuencia relativa en función del total de juegos')
frecuencia_en_funcion_total = dict(zip(["europe", "asia", "africa", "northamerica", "southamerica"], arreglo_frecuencia_en_funcion_total))        
print(frecuencia_en_funcion_total)   
print('\n') 
print('Frecuencia utilizando la lógica del enunciado con np.where')
frecuencia_perdidos = dict(zip(["europe", "asia", "africa", "northamerica", "southamerica"], arreglo_frecuencia_perdidos))        
print(frecuencia_perdidos)


Cantidad de juegos perdidos frecuencia absoluta
{'europe': 14, 'asia': 9, 'africa': 9, 'northamerica': 4, 'southamerica': 3}


Cantidad de juegos
{'europe': 39, 'asia': 12, 'africa': 15, 'northamerica': 12, 'southamerica': 18}


Frecuencia relativa en función del total de juegos
{'europe': 0.358974358974359, 'asia': 0.75, 'africa': 0.6, 'northamerica': 0.3333333333333333, 'southamerica': 0.16666666666666666}


Frecuencia utilizando la lógica del enunciado con np.where
{'europe': 0.6923076923076923, 'asia': 1.0, 'africa': 1.0, 'northamerica': 0.5, 'southamerica': 0.5}


/home/jonathan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


 * ¿Qué continente presentó un nivel similar entre juegos ganados y perdidos?

In [152]:
print('En función de la cantidad total de juegos por frecuencia absoluta')
continentes_arr = ['Europa','Asia','Africa','NorthAmerica','SouthAmerica']
print(continentes_arr)
print(arreglo_frecuencia_absoluta_ganados)
print(arreglo_frecuencia_absoluta_perdidos)
diferencia_total_juegos = [abs(x1 - x2) for (x1, x2) in zip(arreglo_frecuencia_absoluta_ganados, arreglo_frecuencia_absoluta_perdidos)] 
diferencia_ganados_perdidos_total = dict(zip(["europe", "asia", "africa", "northamerica", "southamerica"], diferencia_total_juegos))        
nivel_similar_juegos = sorted(diferencia_ganados_perdidos_total.items(), key=lambda kv: kv[1])
print(nivel_similar_juegos)
print('Por tanto, en función de la cantidad total de juegos a través de la frecuencia absoluta. Norteamérica presentó un nivel similar entre ganados y perdidos')
print('\n')

En función de la cantidad total de juegos por frecuencia absoluta
['Europa', 'Asia', 'Africa', 'NorthAmerica', 'SouthAmerica']
[18, 0, 3, 5, 13]
[14, 9, 9, 4, 3]
[('northamerica', 1), ('europe', 4), ('africa', 6), ('asia', 9), ('southamerica', 10)]
Por tanto, en función de la cantidad total de juegos a través de la frecuencia absoluta. Norteamérica presentó un nivel similar entre ganados y perdidos




* Utilize el código para el caso de juegos perdidos y analice el continente con una mayor frecuencia de juegos perdidos.



Cantidad de juegos perdidos frecuencia absoluta
{'europe': 14, 'asia': 9, 'africa': 9, 'northamerica': 4, 'southamerica': 3}


Cantidad de juegos
{'europe': 39, 'asia': 12, 'africa': 15, 'northamerica': 12, 'southamerica': 18}


Frecuencia relativa en función del total de juegos
{'europe': 0.358974358974359, 'asia': 0.75, 'africa': 0.6, 'northamerica': 0.3333333333333333, 'southamerica': 0.16666666666666666}


Frecuencia utilizando la lógica del enunciado con np.where
{'europe': 0.6923076923076923, 'asia': 1.0, 'africa': 1.0, 'northamerica': 0.5, 'southamerica': 0.5}

**En función del total de juegos el continente con mayor frecuencia relativa de juegos perdidos es Asia**

**Si se observa desde el punto de vista con el planteamiento del uso de la lógica del enunciado a través de np.where los continententes con mayor frecuencia de partidos perdidos serían Asia y Africa**